ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java

```diff
@@ -66,43 +66,52 @@ public CopyUtils(String distCpDoAsUser, HiveConf hiveConf) {
   // Used by replication, copy files from source to destination. It is possible source file is
   // changed/removed during copy, so double check the checksum after copy,
   // if not match, copy again from cm
-  public void copyAndVerify(FileSystem destinationFs, Path destination,
+  public void copyAndVerify(FileSystem destinationFs, Path destRoot,
                     List<ReplChangeManager.FileInfo> srcFiles) throws IOException, LoginException {
-    Map<FileSystem, List<ReplChangeManager.FileInfo>> map = fsToFileMap(srcFiles);
-    for (Map.Entry<FileSystem, List<ReplChangeManager.FileInfo>> entry : map.entrySet()) {
+    Map<FileSystem, Map< Path, List<ReplChangeManager.FileInfo>>> map = fsToFileMap(srcFiles, destRoot);
+    for (Map.Entry<FileSystem, Map<Path, List<ReplChangeManager.FileInfo>>> entry : map.entrySet()) {
       FileSystem sourceFs = entry.getKey();
-      List<ReplChangeManager.FileInfo> fileInfoList = entry.getValue();
-      boolean useRegularCopy = regularCopy(destinationFs, sourceFs, fileInfoList);
+      Map<Path, List<ReplChangeManager.FileInfo>> destMap = entry.getValue();
+      for (Map.Entry<Path, List<ReplChangeManager.FileInfo>> destMapEntry : destMap.entrySet()) {
+        Path destination = destMapEntry.getKey();
+        List<ReplChangeManager.FileInfo> fileInfoList = destMapEntry.getValue();
+        boolean useRegularCopy = regularCopy(destinationFs, sourceFs, fileInfoList);
 
-      doCopyRetry(sourceFs, fileInfoList, destinationFs, destination, useRegularCopy);
-
-      // Verify checksum, retry if checksum changed
-      List<ReplChangeManager.FileInfo> retryFileInfoList = new ArrayList<>();
-      for (ReplChangeManager.FileInfo srcFile : srcFiles) {
-        if(!srcFile.isUseSourcePath()) {
-          // If already use cmpath, nothing we can do here, skip this file
-          continue;
+        if (!destinationFs.exists(destination)
+                && !FileUtils.mkdir(destinationFs, destination, hiveConf)) {
```

Comment:
can mkdir handle the scenario where we need to create nested directory ? I mean the parent directory it self is not present.

Reply:
Yes, HDFS will recursively create the parent directory if not exist.